### PR TITLE
Use VACOLS name for VLJ assignment and validation

### PR DESCRIPTION
**Pull request by odonnadieu**
Resolves https://vajira.max.gov/browse/CASEFLOW-2155

### Description
Updated HearingDaySerializer to serialize first name and last name with the data from CachedUser table. Collect data in batch to avoid any performance hit related to joining of tables for each row. Updated ValidateJudgeSpreadsheet so it validates first and last name with VACOLS data.

### Acceptance Criteria
- [ ] Code compiles correctly
- [ ] Assigning VLJs to dockets with Build Schedule accepts the first and last name from VACOLS in the uploaded file (rather than the full name from Caseflow's users table).
- [ ] Validates that the CSS ID and user name from VACOLS match each other.
- [ ] The Hearings Schedule page and the CSV downloaded from the page populates the "VLJ" column with the name from VACOLS ("<snamel>, <snamef>")

### Testing Plan

1. Go to the console `make c` and set the current user
```ruby
RequestStore[:current_user] = User.system_user
```
2. Create a `User` using `FactoryBot` with the role of judge and a VACOLS record:
```ruby
n = "Leonidas Olson"
id = "BVASLOLSON"
j = FactoryBot.create(:user, :with_vacols_judge_record, full_name: n, css_id: id)
```
3. Associate the newly created user to a `HearingDay` listed in http://localhost:3000/hearings/schedule
```ruby
hd = HearingDay.find(72)
hd.update!(judge: j)
```
4. Find and change the name on the `VACOLS::Staff` record associated with the user: 
```ruby
VACOLS::Staff.find_by(sdomainid: j.css_id).update(snamel: "Gutierrez")
```
5. Cache and sync all the VACOLS users
```ruby
CachedUser.sync_from_vacols
```
6. Shadow user BVASYELLOW http://localhost:3000/test/users
7. Navigate to the hearings schedule http://localhost:3000/hearings/schedule
8. Find the row where the `HearingDay` you associated with the new judge is. The name should match the `VACOLS::Staff` and `CachedUser` records and not the `User` record.
9. Download the table as a CSV by clicking on "Download current view".
10. Open the CSV in excel; edit the sheet so you have 2 or 3 judges. Make sure you have the new `User` you created above (the name here should also match the `VACOLS::Staff` and `CachedUser` records and not the `User` record).
11. Change the name on the sheet to match the name on the `User` table and save the file as an excel worksheet.
12. Navigate to http://localhost:3000/hearings/schedule/build
13. Click on "Upload Files" 
14. Select Judge assignment.
15. Upload the excel sheet you saved above.
16. Click Continue.
17. An error like this should display:
```
These judges are not in the database: [["BVASLOLSON", "Olson, Leonidas"]]
```
18. Edit the name in the spreadsheet to match the name in the `VACOLS::Staff` and `CachedUser` records and save it.
19. Upload the new excel sheet and click continue; you should see the "We have assigned your judges" message and a preview of the changes Caseflow will make on confirmation.

### CODE REVIEW

File: spec/services/hearing_schedule/validate_judge_spreadsheet_spec.rb [hunk 1]
```diff
@@ -1,8 +1,12 @@
 # frozen_string_literal: true
 
 describe HearingSchedule::ValidateJudgeSpreadsheet, :all_dbs do
-  let!(:judge_stuart) { create(:user, full_name: "Stuart Huels", css_id: "BVAHUELS") }
-  let!(:judge_doris) { create(:user, full_name: "Doris Lamphere", css_id: "BVALAMPHERE") }
+  let!(:judge_stuart) { create(:user, :with_vacols_judge_record, full_name: "Stuart Huels", css_id: "BVAHUELS") }
```
Comment (rubaiyat22) [inline spec/services/hearing_schedule/validate_judge_spreadsheet_spec.rb:4]: I'm not sure `judgeNotInDb.xlsx` validates if the name entered does not match our record. It seems to have one row where the css_id value is wrong.  Maybe we could create another row where the css id is `BVALAMPHERE` but it has a wrong name. 

Then on the examples `it "returns JudgeNotInDatabase"` test for the following on top of the current test
```ruby
expect(subject).to include ("['456', 'Huels, Stuart'])
expect(subject).to include ("['BVALAMPHERE', 'wrong first name, wrong last name']")
```

**Review by rubaiyat22 — APPROVED** [review]
This is looking great! I approved but left some comments/questions!

File: app/services/hearing_schedule/assign_judges_to_hearing_days.rb [hunk 2]
```diff
@@ -21,7 +21,7 @@ def stage_assignments(spreadsheet_data)
 
       hearing_days = ::HearingDay.where(id: spreadsheet_data.judge_assignments.pluck(:hearing_day_id))
 
-      hearing_days.map do |hearing_day|
+      hearing_days.sort_by(&:scheduled_for).map do |hearing_day|
```
[the remark this answers is not in the document]
Comment (tomas-nava) [reply app/services/hearing_schedule/assign_judges_to_hearing_days.rb:24]: Not related to the ACs, but I noticed while I was testing that hearing days on the page for confirming judge assignments were not sorted, which made it hard to verify. This makes sure they're sorted by date.

File: app/models/hearing_day.rb [hunk 3]
```diff
@@ -133,7 +135,10 @@ def to_hash
 
     ::HearingDaySerializer.new(
       self,
-      params: { video_hearing_days_request_types: video_hearing_days_request_types }
+      params: {
```
[the remark this answers is not in the document]
Comment (tomas-nava) [reply app/models/hearing_day.rb:138]: Only where we display them in the context of a hearing day

File: app/services/hearing_schedule/validate_judge_spreadsheet.rb [hunk 4]
```diff
@@ -24,18 +24,14 @@ def validate_judge_assignment_template
   def judge_css_id_matches_name?(name, css_id)
```
[the remark this answers is not in the document]
Comment (tomas-nava) [reply app/services/hearing_schedule/validate_judge_spreadsheet.rb]: updated in 16171b19ea9b7d53cf393ba920870ddddd60a1fb

File: spec/services/hearing_schedule/validate_judge_spreadsheet_spec.rb (hunk 1 above)
Comment (tomas-nava) [reply spec/services/hearing_schedule/validate_judge_spreadsheet_spec.rb:4]: good catch! I just replaced the 456 with BVALAMPHERE and updated all the affected tests in d5b259fe710f668020ebe97360d003bdf60c563f

**Pull request merged**